zerver/models/messages.py

```diff
@@ -702,6 +702,29 @@ def __str__(self) -> str:
         return f"{recipient_string} / {self.user_profile.email} ({self.flags_list()})"
 
 
+class IdempotentRequest(models.Model):
+    """Ensure idempotency for Non-Idempotent (e.g. POST) requests.
+
+    Currently, only used for message sending, but in the future it should be applied to
+    any http request having the possibility of being wrongly replayed.
+    """
+
+    idempotency_key = models.UUIDField(unique=True, db_index=True)
```

Comment:
Let's add `realm` and `user` columns to this table, so it can be sharded with the realm, and make the unique index be on `(realm, user, idempotency_key)`. 

That would be more in line with our security model, which isolates realms from each other. And this functionality should isolate users from each other too.